cmd/gardener-admission-controller/app/gardener_admission_controller.go

```diff
@@ -0,0 +1,221 @@
+		return nil, err
+	}
+
+	return o, nil
+}
+
+// loadConfigFromFile loads the contents of file and decodes it as a
+// ControllerManagerConfiguration object.
+func (o *options) loadConfigFromFile(file string) (*config.AdmissionControllerConfiguration, error) {
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+	return o.decodeConfig(data)
+}
+
+func (o *options) configFileSpecified() error {
+	if len(o.ConfigFile) == 0 {
+		return fmt.Errorf("missing Gardener controller manager config file")
+	}
+	return nil
+}
+
+// Validate validates all the required options.
+func (o *options) validate(args []string) error {
+	if len(args) != 0 {
+		return errors.New("arguments are not supported")
+	}
+
+	return nil
+}
+
+// decodeConfig decodes data as a ControllerManagerConfiguration object.
+func (o *options) decodeConfig(data []byte) (*config.AdmissionControllerConfiguration, error) {
+	configObj, gvk, err := o.codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	config, ok := configObj.(*config.AdmissionControllerConfiguration)
+	if !ok {
+		return nil, fmt.Errorf("got unexpected config type: %v", gvk)
+	}
+	return config, nil
+}
+
+func (o *options) applyDefaults(in *config.AdmissionControllerConfiguration) (*config.AdmissionControllerConfiguration, error) {
+	external, err := o.scheme.ConvertToVersion(in, admissioncontrollerconfigv1alpha1.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	o.scheme.Default(external)
+
+	internal, err := o.scheme.ConvertToVersion(external, config.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := internal.(*config.AdmissionControllerConfiguration)
+
+	return out, nil
+}
+
+// AdmissionController contains all necessary information to run the admission controller.
+type AdmissionController struct {
+	Config    *config.AdmissionControllerConfiguration
+	ClientMap clientmap.ClientMap
+}
+
+func (o *options) run(ctx context.Context) error {
+	c, err := o.loadConfigFromFile(o.ConfigFile)
+	if err != nil {
+		return err
+	}
+
+	if c, err = o.applyDefaults(c); err != nil {
+		return err
+	}
+
+	admissionController, err := NewAdmissionController(c)
+	if err != nil {
+		return err
+	}
+
+	return admissionController.Run(ctx)
+}
+
+// NewAdmissionController creates a new AdmissionController instance.
+func NewAdmissionController(cfg *config.AdmissionControllerConfiguration) (*AdmissionController, error) {
+	if cfg == nil {
+		return nil, errors.New("config is required")
+	}
+
+	// Initialize logger
+	logger := logger.NewLogger(cfg.LogLevel)
+	logger.Info("Starting Gardener admission controller...")
+
+	// Prepare a Kubernetes client object for the Garden cluster which contains all the Clientsets
+	// that can be used to access the Kubernetes API.
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		cfg.GardenClientConnection.Kubeconfig = kubeconfig
+	}
```

Comment:
Shouldn't the env var only be used as a fallback, if the config doesn't provide a kubeconfig, instead of overwriting the config?

Reply:
We do this for all of our components this way. Any objection to change it generally @rfranzke ?

Reply:
Let's simply keep it as it is/also was in the past to not break existing usages. It's not an important decision here, either you go left or right, and we decided to go left. :)